src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java

```diff
@@ -0,0 +1,387 @@
+                Eor eor = (Eor) orig;
+                yield g.newEor(orig.getBlock(), eor.getLeft(), eor.getRight());
+            }
+            case iro_Mul -> {
+                Mul mul = (Mul) orig;
+                yield g.newMul(orig.getBlock(), mul.getLeft(), mul.getRight());
+            }
+            case iro_Minus -> {
+                Minus minus = (Minus) orig;
+                yield g.newMinus(orig.getBlock(), minus.getOp());
+            }
+            case iro_Const -> {
+                Const cons = (Const) orig;
+                yield g.newConst(cons.getTarval().asInt(), cons.getMode());
+            }
+            case iro_Sel -> {
+                Sel sel = (Sel) orig;
+                yield g.newSel(orig.getBlock(), sel.getPtr(), sel.getIndex(), sel.getType());
+            }
+            case iro_Not -> {
+                Not not = (Not) orig;
+                yield g.newNot(orig.getBlock(), not.getOp());
+            }
+            case iro_Conv -> {
+                Conv conv = (Conv) orig;
+                yield g.newConv(orig.getBlock(), conv.getOp(), conv.getMode());
+            }
+            default -> null;
+        };
+
+
+        if (newNode != null) {
+            Pointer newPointer = newNode.getBlock().ptr;
+            Pointer origPointer = orig.getBlock().ptr;
+            Pointer replacementPointer = replacement.getBlock().ptr;
+
+            boolean dominatesOld = dominates(newPointer, origPointer);
+            boolean dominatesNew = dominates(newPointer, replacementPointer);
+
+            if (!dominatesOld || !dominatesNew) {
+                // can not replace node if the new node isnt dominating both former nodes.
+                return false;
+            }
+            // new node has to dominate all uses
+            boolean usesDominated = dominatesUses(newNode, orig);
+            boolean usesDominatedRep = dominatesUses(newNode, replacement);
+
+            if (usesDominated && usesDominatedRep) {
+                //      System.out.println("replaced node " + orig + " with " + newNode + " in graph " + g);
+                Graph.exchange(orig, newNode);
+                Graph.exchange(replacement, newNode);
+                return true;
+            } else {
+                return false;
+            }
+        }
+        return false;
+    }
+
+    private boolean hasCallAsChild(Node n) {
+        for (BackEdges.Edge edge : BackEdges.getOuts(n)) {
+            Node node = edge.node;
+            if (node.getOpCode() == binding_irnode.ir_opcode.iro_Call) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private boolean dominates(Pointer source, Pointer target) {
+        int dominates = 0;
+
+        if (!dominateMap.containsKey(source)) {
+            // we have no entry yet, compute both values
+            dominates = binding_irdom.block_dominates(source, target);
+
+            HashMap<Pointer, Integer> cachedResults = new HashMap<>();
```

Comment:
Are you caching for performance reasons only? If so, note that the implementation of `block_dominates` just dereferences a few pointer, and then compares two integers.

Reply:
This is correct, Firm caches the dominance information itself.